### PR TITLE
Avoided unnecessary recompilation of ordering clause regex in SQLCompiler.

### PR DESCRIPTION
**Pull request by adamchainz**
Found this taking up time on every query in a profile.

### CODE REVIEW

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -16,9 +16,15 @@
 from django.db.transaction import TransactionManagementError
 from django.utils.functional import cached_property
 from django.utils.hashable import make_hashable
+from django.utils.regex_helper import _lazy_re_compile
 
 
 class SQLCompiler:
+    ordering_parts = _lazy_re_compile(
+        r'^(.*)\s(ASC|DESC)(.*)',
```
Comment (jdufresne) [inline django/db/models/sql/compiler.py:24]: In my searching, only group 1 is used, so perhaps can avoid unnecessary capturing:

```suggestion
        r'^(.*)\s(?:ASC|DESC).*',
```

Here are the results of the search:

```
django/db/models/sql/compiler.py
381              without_ordering = self.ordering_parts.search(sql).group(1)
394                  without_ordering = self.ordering_parts.search(sql).group(1)
```

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -32,7 +38,6 @@ def __init__(self, query, connection, using):
         self.annotation_col_map = None
         self.klass_info = None
         # Multiline ordering SQL clause may appear from RawSQL.
```
Comment (charettes) [inline django/db/models/sql/compiler.py:40]: I'd move the comment as well.

Comment (adamchainz) [reply django/db/models/sql/compiler.py]: Yes, woops :)

File: django/db/models/sql/compiler.py (hunk 1 above)
Comment (adamchainz) [reply django/db/models/sql/compiler.py]: Well spotted. And very slightly faster

```
In [7]: capturing = re.compile(r'^(.*)\s(ASC|DESC)(.*)')

In [8]: %timeit capturing.search("foo bar ASC baz")
477 ns ± 18.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [9]: non_capturing = re.compile(r'^(.*)\s(?:ASC|DESC).*')

In [10]: %timeit non_capturing.search("foo bar ASC baz")
422 ns ± 1.06 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

**Pull request merged**